### ISSUE TITLE
[FSDP] when inferrence time, the weight changes from 2 batches.

### ISSUE THREAD

**Issue by tani18dr** [facebookresearch/fairscale#545]
## ❓ Questions and Help
hi, I am using fairscale well.

I pretrain and finetune BERT with FullyShardedDataParallel.
However, when inferrence time, the weight changes from 2 batches.

"with torch.no_grad():" and "parameters.requires_grad = False" do not work.

So I did the following.
```
with torch.no_grad():
    model.load_state_dict(torch.load(checkpoint_file))   <---------
    logits = model(input_ids, segment_ids, input_mask)
```
It works well.
Why?
Please help.

**Comment by min-xu-ai** [comment]
very interesting. how do you the weights changed between 2 batches? Can you share a small test that demonstrate this problem?

**Comment by tani18dr** [comment]
I trained with deep speed zero2 and inferred from fp32 with the learned fp16 weights.
The task was carried out by mechine reading comprehension, but in the same way for deep speed, only the first part of the batch in inference was normally answered.

fairscale also showed the same symptoms.

Learning procedure
1. pretrain : korean wiki 3.5 million dataset 50k step zero3 pretrain, two rtx 3090, cuda 11.1 cudnn 8.0.5
2. finetune : mechine reading comprehension 1 epoch zero3 (korquad 1.0 train, https://korquad.github.io/KorQuad%201.0/)
3. evaluate : korquad 1.0 dev 
   -. Loading the model per batch : F1 79.~
   -. one time loading the model : f1 4.~
========================================
    model_chk = torch.load(args.checkpoint, map_location=torch.device(args.device))
    model.load_state_dict(model_chk)
    
    model.eval()
    all_results = []
    logger.info("Start evaluating!")
    for input_ids, input_mask, segment_ids, example_indices in tqdm(dataloader, desc="Evaluating"):
        input_ids = input_ids.to(args.device)
        input_mask = input_mask.to(args.device)
        segment_ids = segment_ids.to(args.device)
        with torch.no_grad():
            if args.zero:
                model.load_state_dict(model_chk)
            batch_start_logits, batch_end_logits = model(
                input_ids = input_ids,
                token_type_ids = segment_ids, 
                attention_mask = input_mask,
            )
        for i, example_index in enumerate(example_indices):
            start_logits = batch_start_logits[i].detach().cpu().tolist()
            end_logits = batch_end_logits[i].detach().cpu().tolist()
            eval_feature = eval_features[example_index.item()]
            unique_id = int(eval_feature.unique_id)
            all_results.append(RawResult(unique_id=unique_id,
                                         start_logits=start_logits,
                                         end_logits=end_logits))
    output_prediction_file = os.path.join(args.output_dir, "predictions.json")
    output_nbest_file = os.path.join(args.output_dir, "nbest_predictions.json")

    write_predictions(eval_examples, eval_features, all_results,
                  args.n_best_size, args.max_answer_length,
                  False, output_prediction_file, output_nbest_file,
                  None, False, False, 0.0)

**Comment by min-xu-ai** [comment]
by "symptom", do you mean the two scores are different in the following?

> -. Loading the model per batch : F1 79.~
> -. one time loading the model : f1 4.~

I think I understand the problem better now. Thanks for the additional info. The additional line of `model.load_state_dict(model_chk)` after `if args.zero` seems to help your eval score. This shouldn't be happening. Let me see if I can reproduce it or not.

**Comment by min-xu-ai** [comment]
Sorry for the delay. I have not been able to reproduce. Is this still happening for you? Could you possibly create a small test case where this problem is happening?

**Comment by tani18dr** [comment]
Thank you.
I don't know why, but I can't reproduce the change from version 0.3.2 to version 0.3.6.
Now I am using fairscale well. 
Thanks again.

**Comment by min-xu-ai** [comment]
Thank you so much!
